framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -896,31 +910,63 @@ protected String dotOutputFileName(UnderlyingAST ast) {
         return null;
     }
 
+    /**
+     * Returns the type of the left-hand side of an assignment without
+     * applying local variable defaults to type variables.
+     *
+     * The type variables that are types of local variables are defaulted to
```

Comment:
In your code example, `T`'s lower bound is `@NonNull Void` and its upper bound is `@Nullable Object`.  `@Nullable T`'s lower bound is `@Nullable Void` and its upper bound is `@Nullable Object`. So `@Nullable T` is a super type of all possible values of `T`.  (And this is true whatever the annotations are on the declaration of `T`.)

So the refined type of `local` after the assignment is `T`  ( with lower bound `@NonNull Void` and upper bound `@Nullable Object`), so the return is legal.

Type variables are treated differently during type argument inference.  For example:
  ````
<S> S bar () {...}
<T> T foo(T p) {
  T local = bar();
  return local;
}
````
During type argument inference of `bar`, the assignment context is  `local`.  If the local variable default is used, then the type of assignment context type is `@Nullable T` and the type argument inferred for `bar()` is `@Nullable T`.  And an incompatible types in return error is issued.

If instead, the local variable default is not applied, then the assignment context type is `T` (with lower bound `@NonNull Void` and upper bound `@Nullable Object`) and the type argument inferred for `bar()` is `T`.  During dataflow, the type of `local` is refined to `T` and the return is legal. 

If the assignment context type was a declared type, for example:
  ````
<S> S bar () {...}
Object foo() {
   Object local = bar();
  return local;
}
````
The local variable default must be used or else the assignment context type is missing an annotation.  So, an incompatible types in return error is issued in the above code.  We could improve type argument inference in this case and by using the lower bound of `S` instead of the local variable default.